util/dvsim/NcLauncher.py

```diff
@@ -0,0 +1,214 @@
+            f'cat post.log >>{log_file}'
+        )
+
+        # TODO: These tool-specific names need moving into an hjson config
+        # file.
+        if (exetool == 'xcelium'):
+            license_args = ['-r', 'License:Xcelium_Single_Core/1']
+        elif (exetool == 'vcs'):
+            license_args = ['-r', 'License:VCSRuntime_Net/1']
+        else:
+            license_args = []
+
+        self.create_run_sh(odir, cmd)
+
+        return (['nc', 'run',
+                 '-e', 'SNAPSHOT',
+                 '-nodb', '-forcelog',
+                 '-l', log_file,
+                 '-postcmd', postcmd,
+                 '-set', job_name] +
+                license_args +
+                interactive_flags +
+                ['--', f'{odir}/run.sh'])
+
+    def _do_launch(self):
+        # Compute the environment for the subprocess by overriding environment
+        # variables of this process with matching ones from self.deploy.exports
+        exports = os.environ.copy()
+        if self.deploy.exports:
+            exports.update(self.deploy.exports)
+
+        # Clear the magic MAKEFLAGS variable from exports if necessary. This
+        # variable is used by recursive Make calls to pass variables from one
+        # level to the next. Here, self.cmd is a call to Make but it's
+        # logically a top-level invocation: we don't want to pollute the flow's
+        # Makefile with Make variables from any wrapper that called dvsim.
+        if 'MAKEFLAGS' in exports:
+            del exports['MAKEFLAGS']
+
+        self._dump_env_vars(exports)
+
+        if self.deploy.sim_cfg.interactive:
+            # Interactive: Set RUN_INTERACTIVE to 1
+            exports['RUN_INTERACTIVE'] = '1'
+            interactive_flags = ['-I']
+            cmd_arr = self.get_submit_cmd(interactive_flags)
+            log.log(VERBOSE, '[Executing]:\n{}\n\n'.format(self.deploy.cmd))
+            self.process = subprocess.Popen(cmd_arr,
+                                            stdin=None,
+                                            stdout=None,
+                                            stderr=subprocess.STDOUT,
+                                            # string mode
+                                            universal_newlines=True,
+                                            env=exports,
+                                            cwd=self.deploy.odir)
+
+            # Wait until the process exits
+            self.process.wait()
+        else:
+            try:
+                interactive_flags = ['-nolog', '-wl']
+                cmd_arr = self.get_submit_cmd(interactive_flags)
+                with open(self.deploy.get_log_path(),
+                          'w',
+                          encoding='UTF-8',
+                          errors='surrogateescape') as f:
+                    f.write('[Executing]:\n{}\n\n'.format(self.deploy.cmd))
+                    self.process = subprocess.Popen(cmd_arr,
```

Comment:
Oof: this looks a little surprising to me (which I think might have been my fault!) Does't `Popen` exit almost immediately? The `with` block will then finish. Presumably Python avoids closing the fds because they are pointed to by `self.process`.

My suggestion was terrible! Probably best to keep the explicit `open` call that you had before (possibly with a comment to avoid confusing the next person like me...). I'm really sorry for the noise.

Reply:
done . reverted the changes with a comment explaining the reason